codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,467 @@
+                unexpected = true;
+            }
+            if (unexpected) {
+                // TODO: swap to failing the promise. Today, however, the promise is likely being
+                // ignored.
+                throw new IllegalStateException("Unexpected Http2Frame type: " + msg.getClass().getName());
+            }
+        } finally {
+            ReferenceCountUtil.release(msg);
+        }
+    }
+
+    Http2StreamChannel createStreamChannel(ChannelHandlerContext ctx, Http2Stream stream,
+            ChannelHandler handler) {
+        EventLoopGroup group = streamGroup != null ? streamGroup : ctx.channel().eventLoop();
+        Http2StreamChannel channel = new Http2StreamChannel(stream);
+        channel.pipeline().addLast(handler);
+        group.register(channel);
+        return channel;
+    }
+
+    private static ChannelHandlerInvoker invoker(ChannelHandlerContext ctx) {
+        return ((EventLoop) ctx.executor()).asInvoker();
+    }
+
+    /**
+     * Notifies any child streams of the read completion.
+     */
+    @Override
+    public void channelReadComplete(ChannelHandlerContext ctx) {
+        for (int i = 0; i < streamsToFireChildReadComplete.size(); i++) {
+            final StreamInfo streamInfo = streamsToFireChildReadComplete.get(i);
+            // Clear early in case fireChildReadComplete() causes it to need to be re-processed
+            streamInfo.inStreamsToFireChildReadComplete = false;
+            streamInfo.childChannel.fireChildReadComplete();
+        }
+        streamsToFireChildReadComplete.clear();
+    }
+
+    void fireChildReadAndRegister(StreamInfo streamInfo, Object msg) {
+        // Can't use childChannel.fireChannelRead() as it would fire independent of whether
+        // channel.read() had been called.
+        streamInfo.childChannel.fireChildRead(msg);
+        if (!streamInfo.inStreamsToFireChildReadComplete) {
+            streamsToFireChildReadComplete.add(streamInfo);
+            streamInfo.inStreamsToFireChildReadComplete = true;
+        }
+    }
+
+    final class ConnectionListener extends Http2ConnectionAdapter {
+        @Override
+        public void onStreamActive(Http2Stream stream) {
+            if (ctx == null) {
+                // UPGRADE stream is active before handlerAdded().
+                return;
+            }
+            // If it is an outgoing stream, then we already created the channel.
+            // TODO: support outgoing streams
+            if (stream.getProperty(streamInfoKey) != null) {
+                return;
+            }
+            stream.setProperty(streamInfoKey,
+                new StreamInfo(createStreamChannel(ctx, stream, streamHandler)));
+        }
+
+        @Override
+        public void onStreamClosed(Http2Stream stream) {
+            final StreamInfo streamInfo = (StreamInfo) stream.removeProperty(streamInfoKey);
+            if (streamInfo != null) {
+                if (streamInfo.childChannel.eventLoop().inEventLoop()) {
```

Comment:
`eventLoop()` is volatile under the hood. Consider saving a reference to it so we only have 1 volatile read.